more TLS cleanup.

jboeuf opened the pull request:
- using NULL-terminated strings as opposed to const unsigned char *
with length since the credentials are in PEM format.
- aligning the structures with gRPC so that we have less convertions to
do back and forth.

grpc-jenkins commented:
```
No significant performance differences
```

markdroth commented:
I'm not super familiar with the SSL code, but this looks like a nice simplification.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/core/lib/security/credentials/ssl/ssl_credentials.c, line 53 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-Khbkp1azepOC7qreKog:-Khbkp1azepOC7qreKoh:bmwyf5r) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/credentials/ssl/ssl_credentials.c#L53)):*
> ```C
>     tsi_ssl_pem_key_cert_pair *kp) {
>   if (kp == NULL) return;
>   if (kp->private_key != NULL) gpr_free((void *)kp->private_key);
> ```

There's no need to check if the value is NULL before calling gpr_free().  That function is a no-op if the argument is NULL.

Same thing throughout.

---

*[src/core/lib/security/credentials/ssl/ssl_credentials.c, line 106 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhbkE5moS0YKiQvX9Lw:-KhbkE5moS0YKiQvX9Lx:bkqd3vt) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/credentials/ssl/ssl_credentials.c#L106)):*
> ```C
>     config->pem_root_certs = gpr_strdup(pem_root_certs);
>   }
>   if (pem_key_cert_pair != NULL) {
> ```

Do we also need to set `config->has_key_cert_pair = true`?  (But see my comment elsewhere about removing this field.)

---

*[src/core/lib/security/transport/security_connector.c, line 828 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-Khbiau5s6yjGZNCcLWJ:-Khbiau5s6yjGZNCcLWK:bh79br6) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/transport/security_connector.c#L828)):*
>   tsi_ssl_pem_key_cert_pair cert_key_pair = {
>       config->pem_key_cert_pair.private_key,
>       config->pem_key_cert_pair.cert_chain};
>  
>   result = tsi_create_ssl_client_handshaker_factory(
>       config->has_key_cert_pair ? &cert_key_pair : NULL,

Why define a local `tsi_ssl_pem_key_cert_pair` instance here?  Why not just pass `&config->pem_key_cert_pair` to `tsi_create_ssl_client_handshaker_factory()`?

---

*[src/core/lib/security/transport/security_connector.h, line 193 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-Khbirm534JBlBqPdMUH:-Khbirm534JBlBqPdMUI:bs160n7) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/transport/security_connector.h#L193)):*
> ```objc
> typedef struct {
>   tsi_ssl_pem_key_cert_pair pem_key_cert_pair;
>   bool has_key_cert_pair;
> ```

Do we actually need this bool?  Could we instead check that the fields of `pem_key_cert_pair` are non-NULL?

---

*[src/core/tsi/ssl_transport_security.c, line 691 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhbgXUHnH6Rrw-tcAvR:-KhbgXUHnH6Rrw-tcAvS:buu7y4z) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/tsi/ssl_transport_security.c#L691)):*
> ```C
>   for (i = 0; i < num_alpn_protocols; i++) {
>     size_t length = alpn_protocols[i] == NULL ? 0 : strlen(alpn_protocols[i]);
>     if (length == 0 || length > 255) {
> ```

Where does the limit of 255 characters come from?  Doesn't look like the code was checking this before.

---

*[src/core/tsi/ssl_transport_security.h, line 75 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhbfhOJiZA1FdrxMyun:-KhbfhOKl9Fr4vlIOOaO:b-our5dn) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/tsi/ssl_transport_security.h#L75)):*
> ```objc
> 
> /* Creates a client handshaker factory.
>    - pem_key_cert_pair is a pointer on the object containing client's private
> ```

s/on/to/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10627#-:-KhblIUGAura5K3Eza-l:b-2i3s64)*
<!-- Sent from Reviewable.io -->

jboeuf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/core/lib/security/credentials/ssl/ssl_credentials.c, line 106 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhbkE5moS0YKiQvX9Lw:-KhcqXdKC79jgepBuv0r:b-1qdqe3) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/credentials/ssl/ssl_credentials.c#L106)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Do we also need to set `config->has_key_cert_pair = true`?  (But see my comment elsewhere about removing this field.)
</blockquote></details>

good point. removed the bool in the config.

---

*[src/core/lib/security/transport/security_connector.c, line 828 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-Khbiau5s6yjGZNCcLWJ:-KhcqcxzG22N-rULboIO:b-esfere) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/transport/security_connector.c#L828)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

>   tsi_ssl_pem_key_cert_pair cert_key_pair = {
>       config->pem_key_cert_pair.private_key,
>       config->pem_key_cert_pair.cert_chain};
>  
>   result = tsi_create_ssl_client_handshaker_factory(
>       config->has_key_cert_pair ? &cert_key_pair : NULL,

Why define a local `tsi_ssl_pem_key_cert_pair` instance here?  Why not just pass `&config->pem_key_cert_pair` to `tsi_create_ssl_client_handshaker_factory()`?
</blockquote></details>

left-over from a previous version of the code. good catch.

---

*[src/core/tsi/ssl_transport_security.c, line 691 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhbgXUHnH6Rrw-tcAvR:-KhcrDFJSJ-KHzBVtAiY:b-uy4z8q) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/tsi/ssl_transport_security.c#L691)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Where does the limit of 255 characters come from?  Doesn't look like the code was checking this before.
</blockquote></details>

The previous code had the length precomputed in an array. The new code just uses strlen and since the encoding of the length is on 1 byte, we have to check that the length will not overflow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10627)*
<!-- Sent from Reviewable.io -->

jboeuf commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[src/core/lib/security/credentials/ssl/ssl_credentials.c, line 53 at r1](https://reviewable.io:443/reviews/grpc/grpc/10627#-Khbkp1azepOC7qreKog:-Khd3o0KXT32gx0CJay8:b-7rl0jd) ([raw file](https://github.com/grpc/grpc/blob/5244bdf1cc78151e855f85c2d0139ffbec7a5a3d/src/core/lib/security/credentials/ssl/ssl_credentials.c#L53)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

There's no need to check if the value is NULL before calling gpr_free().  That function is a no-op if the argument is NULL.

Same thing throughout.
</blockquote></details>

done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10627)*
<!-- Sent from Reviewable.io -->

jboeuf commented:
Thanks Mark. I addressed all the comments. 

Regarding the length check at 255 chars:
The previous code had the length precomputed in an array. The new code just uses strlen and since the encoding of the length is on 1 byte, we have to check that the length will not overflow.

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                  cpu_time    real_time
-------------------------------------------------------  ----------  -----------
BM_HpackParserParseHeader<AddIndexedSingleInternedElem>      +40.00       +40.00
```

markdroth commented:
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/core/lib/security/credentials/ssl/ssl_credentials.c, line 144 at r2](https://reviewable.io:443/reviews/grpc/grpc/10627#-KhdMRbDM_IKikRtRRyw:-KhdMRbEH4sq3gFySioX:b-nrr5n0) ([raw file](https://github.com/grpc/grpc/blob/3a25149314a1f1ab78a171edc606231f5487b853/src/core/lib/security/credentials/ssl/ssl_credentials.c#L144)):*
>   if (c->config.pem_key_cert_pairs != NULL)
>     gpr_free(c->config.pem_key_cert_pairs);
>   if (c->config.pem_root_certs != NULL) gpr_free(c->config.pem_root_certs);

No need to check if NULL before freeing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/10627)*
<!-- Sent from Reviewable.io -->

grpc-jenkins commented:
```
Performance differences noted:
Benchmark                                                                cpu_time    real_time
---------------------------------------------------------------------  ----------  -----------
BM_HpackEncoderEncodeHeader<SingleInternedBinaryElem<10, true>>/0/16k       +7.50        +7.50
```

grpc-jenkins commented:
```
No significant performance differences
```

jboeuf commented:
@dgquintas are you comfortable with me merging this PR?

dgquintas commented:
There are a lot of interop failures (https://grpc-testing.appspot.com/job/gRPC_interop_pull_requests/13483/).

grpc-jenkins commented:
```
No significant performance differences
```

jboeuf commented:
interop tests are green after rebasing with master.

dgquintas commented:
There's only one failure that may be related to the PR: https://grpc-testing.appspot.com/job/gRPC_pull_requests_asan_c/4082/testReport/junit/(root)/c_linux_asan_native/bins_asan_h2_sockpair_1byte_test_graceful_server_shutdown_GRPC_POLL_STRATEGY_poll_cv/

For the rest, can you merge with master? Some failures have been fixed since, and others were due to failure to clone submodules and such.

grpc-testing commented:
```
No significant performance differences
```

jboeuf commented:
This test failure is not there anymore. There are others though but they look unrelated. I'm merging now, please revert if you see an issue. Thanks!
